Improve StringTypeAdapter templates and handling

zdobersek opened the pull request:
#### 4b89b603e1e19e17fed1c93f72872821b4a2b984
<pre>
Improve StringTypeAdapter templates and handling
<a href="https://bugs.webkit.org/show_bug.cgi?id=250017">https://bugs.webkit.org/show_bug.cgi?id=250017</a>

Reviewed by NOBODY (OOPS!).

StringConcatenate&apos;s makeString() and makeAtomString() are adjusted to accept
arguments as const lvalue references, propagating them down to each
StringTypeAdapter object without unnecessary copies. Similarly, the adapter
objects too are handled through const lvalue references once constructed and
passed on.

Newly-imposed constness requires additional const qualifiers on common methods
across different StringTypeAdapter specializations.

The are8Bit() and stringTypeAdapterAccumulator() helper functions are simplified
through using parameter packs and fold expressions to perform the necessary
operations across a range of adapter objects.

StringTypeAdapter specialization for StringView is moved from the StringView
header into the StringConcatenate header. A new partial specialization for the
StringTypeAdapter template is added, conditionally enabled if a const reference
of the specified type can be used to construct a StringView object. This
specialization inherits from the StringView specialization, hence a StringView
is constructed for such objects.

An exception for this partial specialization are character string literals whose
type is reduced to a fixed-length C array. These for the moment require a
separate specialization in order to avoid implicitly creating a String object
that&apos;s then used for the StringView. The fixed-length array data is instead
passed to the partial specializations handling string pointers, in order to
pass the data through string-length computation and getting the correct length.
This avoids potential problems with fixed-length arrays with inflated size that
are used as buffers for string data that&apos;s otherwise null-terminated at some
previous position. Unit tests are provided to cover this detail.

Related to this, additional constructors on the String and StringView classes
accepting fixed-length C array values are deleted. This should prevent
constructing String and StringView objects from string literals or fixed-length
C arrays by using implicit conversion through StaticStringImpl or Span objects.

This enables removing a bunch of other StringTypeAdapter specializations.
Different WTF string and impl types can be used to construct the StringView,
as well as ASCIILiteral. For Vectors of character values, a corresponding Span
can be implicitly constructed and is then used for StringView construction.

In StringView, constructors taking in ASCIILiteral or Span are adjusted to
handle const lvalue references of those objects.

Different static integer values can now end up generating undesired symbols when
handled through references in calls to makeString() and makeAtomString(). This
can be avoided by explicitly copying the passed-in value.

* Source/JavaScriptCore/inspector/InjectedScriptBase.cpp:
(Inspector::InjectedScriptBase::makeCall):
(Inspector::InjectedScriptBase::makeAsyncCall):
* Source/WTF/wtf/text/StringConcatenate.h:
(WTF::are8Bit):
(WTF::stringTypeAdapterAccumulator):
(WTF::tryMakeStringImplFromAdaptersInternal):
(WTF::handleWithAdapters):
(WTF::tryMakeStringFromAdapters):
(WTF::tryMakeString):
(WTF::makeString):
(WTF::tryMakeAtomStringFromAdapters):
(WTF::tryMakeAtomString):
(WTF::makeAtomString):
* Source/WTF/wtf/text/StringImpl.h:
* Source/WTF/wtf/text/StringOperators.h:
(WTF::StringAppend::is8Bit const):
(WTF::StringAppend::writeTo const):
(WTF::StringAppend::length const):
(WTF::StringAppend::is8Bit): Deleted.
(WTF::StringAppend::writeTo): Deleted.
(WTF::StringAppend::length): Deleted.
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::StringView):
* Source/WTF/wtf/text/WTFString.h:
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::generateTemporaryPath):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/dom/make_names.pl:
(printTagNameCppFile):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:
* Tools/TestWebKitAPI/Tests/WTF/StringConcatenate.cpp:
(TestWebKitAPI::TEST):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b89b603e1e19e17fed1c93f72872821b4a2b984

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105196 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14279 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38079 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114455 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174635 "Hash 4b89b603 for PR 8712 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109104 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15427 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5196 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97510 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114397 "Hash 4b89b603 for PR 8712 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110952 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11924 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94914 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39422 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93792 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26549 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81088 "Found 3 new API test failures: /TestWTF:WTF_ParkingLot.UnparkOneOneFast, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state, /WebKitGTK/TestDownloads:/webkit/Downloads/local-file-error (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/95017 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7610 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27908 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/93078 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5340 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7705 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4481 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30320 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13757 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47463 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101780 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9493 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25398 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WTF/wtf/text/StringConcatenate.h:
```diff
@@ -148,104 +148,38 @@ template<> class StringTypeAdapter<char*, void> : public StringTypeAdapter<const
 };
 
-template<> class StringTypeAdapter<AtomStringImpl&, void> : public StringTypeAdapter<StringImpl&, void> {
+// Array types in the following partial specialization are disallowed in order to avoid implicit constructors
```
Which String implicit ctor is used?
Does String have an implicit ctor for a character array?

fujii replied:
UserAgentGLib.cpp is compiled only by GTK and WPE. Let's add a test case in https://github.com/WebKit/WebKit/blob/main/Tools/TestWebKitAPI/Tests/WTF/StringConcatenate.cpp .

zdobersek replied:
> Which String implicit ctor is used?
> Does String have an implicit ctor for a character array?

Via `StaticStringImpl`.

fujii replied:
We have `""_s` operator. I think we should disallow constructing a String from a character array. Right?

fujii replied:
I looked into this. Interesting problem. But, I confirmed that StringView can't be constructed from a string literal. I think the usage of is_constructible_v can be impoved. How about thi patch?
https://bugs.webkit.org/show_bug.cgi?id=250017#c10

zdobersek replied:
> We have `""_s` operator. I think we should disallow constructing a String from a character array. Right?

I would say yes, even if it's a pretty big restriction. I imagine it happens a lot in the code.

zdobersek replied:
> I looked into this. Interesting problem. But, I confirmed that StringView can't be constructed from a string literal. I think the usage of is_constructible_v can be impoved. How about thi patch?
> https://bugs.webkit.org/show_bug.cgi?id=250017#c10

Thanks for the proposal, I'll investigate it.

I have a few other changes queued up behind this one, including again supporting string literals in StringView, but this time with enforced length computation (i.e. no more dependence on the array size value). I'll notify you on relevant PRs as I open them.

fujii replied:
> including again supporting string literals in StringView

Is it a good idea? Why do we have ASCIILiteral?  StringView explicitly disallows it.
https://github.com/WebKit/WebKit/blob/1b38f29b8d045a32314964794baa0094954e3b69/Source/WTF/wtf/text/StringView.h#L198-L200

zdobersek replied:
ASCIILiteral for sure is preferred, but support for string literals (after properly handling the string length) is cheap and would also cover UChar literals (or rather arrays of UChars). But usage of those isn't really common, and for the purposes of `makeString()` and friends it can be handled just fine through the `StringTypeAdapter` specialization.

Initially I will actually be proposing deleting `String` and `StringView` constructors that would take in `CharacterType[N]` literals, to disallow implicit specialization with UChar data not just through `StaticStringImpl` but apparently also through one implicit constructor of `Span<const UChar>`.

fujii replied:
It sounds nice, but it's against https://github.com/WebKit/WebKit/pull/1040 .
Let's make a consensus rather than edit warring.

zdobersek replied:
OK. I agree supporting raw string literals (i.e. fixed-length LChar or UChar arrays) isn't the best option, even if it's trivial to support. I can add them in the future if there's actual benefits with them being present.

fujii replied:
I asked them. This is the reason why we should disallow `const char*` for StringView and makeString.
https://bugs.webkit.org/show_bug.cgi?id=240942#c11

zdobersek committed:
Improve StringTypeAdapter templates and handling
https://bugs.webkit.org/show_bug.cgi?id=250017

Reviewed by NOBODY (OOPS!).

StringConcatenate's makeString() and makeAtomString() are adjusted to accept
arguments as const lvalue references, propagating them down to each
StringTypeAdapter object without unnecessary copies. Similarly, the adapter
objects too are handled through const lvalue references once constructed and
passed on.

Newly-imposed constness requires additional const qualifiers on common methods
across different StringTypeAdapter specializations.

The are8Bit() and stringTypeAdapterAccumulator() helper functions are simplified
through using parameter packs and fold expressions to perform the necessary
operations across a range of adapter objects.

StringTypeAdapter specialization for StringView is moved from the StringView
header into the StringConcatenate header. A new partial specialization for the
StringTypeAdapter template is added, conditionally enabled if a const reference
of the specified type can be used to construct a StringView object. This
specialization inherits from the StringView specialization, hence a StringView
is constructed for such objects.

An exception for this partial specialization are character string literals whose
type is reduced to a fixed-length C array. These for the moment require a
separate specialization in order to avoid implicitly creating a String object
that's then used for the StringView. The fixed-length array data is instead
passed to the partial specializations handling string pointers, in order to
pass the data through string-length computation and getting the correct length.
This avoids potential problems with fixed-length arrays with inflated size that
are used as buffers for string data that's otherwise null-terminated at some
previous position. Unit tests are provided to cover this detail.

Related to this, additional constructors on the String and StringView classes
accepting fixed-length C array values are deleted. This should prevent
constructing String and StringView objects from string literals or fixed-length
C arrays by using implicit conversion through StaticStringImpl or Span objects.

This enables removing a bunch of other StringTypeAdapter specializations.
Different WTF string and impl types can be used to construct the StringView,
as well as ASCIILiteral. For Vectors of character values, a corresponding Span
can be implicitly constructed and is then used for StringView construction.

In StringView, constructors taking in ASCIILiteral or Span are adjusted to
handle const lvalue references of those objects.

Different static integer values can now end up generating undesired symbols when
handled through references in calls to makeString() and makeAtomString(). This
can be avoided by explicitly copying the passed-in value.

* Source/JavaScriptCore/inspector/InjectedScriptBase.cpp:
(Inspector::InjectedScriptBase::makeCall):
(Inspector::InjectedScriptBase::makeAsyncCall):
* Source/WTF/wtf/text/StringConcatenate.h:
(WTF::are8Bit):
(WTF::stringTypeAdapterAccumulator):
(WTF::tryMakeStringImplFromAdaptersInternal):
(WTF::handleWithAdapters):
(WTF::tryMakeStringFromAdapters):
(WTF::tryMakeString):
(WTF::makeString):
(WTF::tryMakeAtomStringFromAdapters):
(WTF::tryMakeAtomString):
(WTF::makeAtomString):
* Source/WTF/wtf/text/StringImpl.h:
* Source/WTF/wtf/text/StringOperators.h:
(WTF::StringAppend::is8Bit const):
(WTF::StringAppend::writeTo const):
(WTF::StringAppend::length const):
(WTF::StringAppend::is8Bit): Deleted.
(WTF::StringAppend::writeTo): Deleted.
(WTF::StringAppend::length): Deleted.
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::StringView):
* Source/WTF/wtf/text/WTFString.h:
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::generateTemporaryPath):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/dom/make_names.pl:
(printTagNameCppFile):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:
* Tools/TestWebKitAPI/Tests/WTF/StringConcatenate.cpp:
(TestWebKitAPI::TEST):

fujii commented on Source/WTF/wtf/win/FileSystemWin.cpp:
```diff
@@ -224,7 +224,7 @@ static String generateTemporaryPath(const Function<bool(const String&)>& action)
         ASSERT(wcslen(tempFile) == std::size(tempFile) - 1);
 
-        proposedPath = pathByAppendingComponent(String(tempPath), String(tempFile));
+        proposedPath = pathByAppendingComponent(String(std::data(tempPath)), String(std::data(tempFile)));
```
Using std::data doesn't look correct to me. It makes me feeling something wrong.

fujii replied:
I came up with a good idea to solve this problem. Can I take over this task?

zdobersek replied:
I spun off the constructor deletion and relevant fixes to #9282. You can continue there.

`std::data()` just converts the fixed-length array value into the corresponding pointer value, avoiding the deleted constructors.
